### PR TITLE
Migrate from Cassandra to PostgreSQL

### PR DESCRIPTION
**Pull request by Guaris**
https://konghq.atlassian.net/browse/DOCU-3299

Migration guidelines from Cassandra to PostgreSQL

### CODE REVIEW

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+---
+title: Migration guidelines: Cassandra to Postgres
+---
+
+The purpose of this document is to provide an approach aimed at migrating Cassandra db backed Kong deployment to Postgresql backed Kong deployment. This document will guide migration from traditional Kong deployment topology to hybrid.
+
+A Blue/Green migration approach is proposed in this document. To achieve the easiest possible migration, declarative tools Kong [DecK](https://docs.konghq.com/deck/latest/) will be utilized. DNS switching will be used to cut traffic over from a Blue environment to a Green environment. This will provide the ability to rollback faster if an issue is detected. 
+
+The steps to migrate, at a high level are:
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:9]:
```suggestion
Migration steps:
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+title: Migration guidelines: Cassandra to Postgres
+---
+
+The purpose of this document is to provide an approach aimed at migrating Cassandra db backed Kong deployment to Postgresql backed Kong deployment. This document will guide migration from traditional Kong deployment topology to hybrid.
+
+A Blue/Green migration approach is proposed in this document. To achieve the easiest possible migration, declarative tools Kong [DecK](https://docs.konghq.com/deck/latest/) will be utilized. DNS switching will be used to cut traffic over from a Blue environment to a Green environment. This will provide the ability to rollback faster if an issue is detected. 
+
+The steps to migrate, at a high level are: 
+
+* New target Kong environment will be built as a blank canvas with Postgresql db with hybrid deployment.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:11]:
```suggestion
1. Build a target {{site.base_gateway}} hybrid environment with PostgreSQL. This acts as your blank canvas.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+---
+
+The purpose of this document is to provide an approach aimed at migrating Cassandra db backed Kong deployment to Postgresql backed Kong deployment. This document will guide migration from traditional Kong deployment topology to hybrid.
+
+A Blue/Green migration approach is proposed in this document. To achieve the easiest possible migration, declarative tools Kong [DecK](https://docs.konghq.com/deck/latest/) will be utilized. DNS switching will be used to cut traffic over from a Blue environment to a Green environment. This will provide the ability to rollback faster if an issue is detected. 
+
+The steps to migrate, at a high level are: 
+
+* New target Kong environment will be built as a blank canvas with Postgresql db with hybrid deployment. 
+* Kong deck is used to take export (dump) from old traditional cassandra backed Kong deployment.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:12]:
```suggestion
1. Use decK to export (`dump`) configuration from the old traditional Cassandra-backed deployment. 
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+The purpose of this document is to provide an approach aimed at migrating Cassandra db backed Kong deployment to Postgresql backed Kong deployment. This document will guide migration from traditional Kong deployment topology to hybrid.
+
+A Blue/Green migration approach is proposed in this document. To achieve the easiest possible migration, declarative tools Kong [DecK](https://docs.konghq.com/deck/latest/) will be utilized. DNS switching will be used to cut traffic over from a Blue environment to a Green environment. This will provide the ability to rollback faster if an issue is detected. 
+
+The steps to migrate, at a high level are: 
+
+* New target Kong environment will be built as a blank canvas with Postgresql db with hybrid deployment. 
+* Kong deck is used to take export (dump) from old traditional cassandra backed Kong deployment. 
+* Kong deck is used to import (sync) to new hybrid postgresql backed Kong deployment.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:13]:
```suggestion
1. Use decK to import (`sync`) the configuration to the new hybrid PostgreSQL-backed deployment.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+* Basic auth credentials and local rbac users will need to be created separately either using admin api or manually(Kong Manager).
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:14]:
```suggestion
1. Create basic auth credentials and local RBAC users using either the Admin Api or Kong Manager.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+* When the Kong configuration has been migrated over to the green environment, traffic can be slowly canaried into the green environment to de-risk the release with a fast rollback mechanism in place.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:15]:
```suggestion
1. When the {{site.base_gateway}} configuration has been migrated over to the green environment, slowly redirect traffic via canary into the green environment. This lowers the risk of the release with a fast rollback mechanism in place.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+
+### Considerations
+
+While the document provides high level guidelines, actual migration steps may differ based on the following factors:
+* complexity of the Kong deployment environment
+* Kong plugins used
+* Usage of custom plugins
+* External touchpoints to the Gateway (e.g. OIDC)
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:24]:
```suggestion
* Complexity of the Kong deployment environment
* Kong plugins
* Custom plugins
* External touchpoints to the Gateway (for example, OIDC with external IdPs)
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+* Using cassandra means may not be using the Kong Hybrid Mode deployment and might be deployed in the traditional mode, we recommend taking this opportunity to review the architecture to leverage the latest features of Kong Enterprise.
+
+While the diagram shows the split between the Kong control and Data planes, the same approach can be followed for Kong instances deployed in Classic Mode.
+
+![migration image](_assets/images/docs/gateway/migration.png)
+
+
+### Assumptions
+The Kong blue environment (Using Cassandra) and green environment (using Postgres) are running the same Kong Enterprise version
+Blue is deployed in traditional mode and green environment is running with Kong deployed in a hybrid deployment mode.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:36]:
```suggestion
## Prerequisites
* The {{site.base_gateway}} blue environment (using Cassandra) and green environment (using PostgreSQL) are running the same Gateway version.
* In the following examples, blue is deployed in traditional mode and green is deployed in hybrid mode. If you prefer to migrate to a traditional deployment, you can still follow the steps, but you don't need to worry about performing separate control plane and data plane tasks.
```
This guide should work for either OSS or Enterprise, correct?

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+
+## Migration Approach
+The following steps should be tested in a non-production environment. Any gaps in the target state should be identified and remediated before running it in production.
+
+| Step | Name               | Description                                                                                           | Completed |
+|------|--------------------|-------------------------------------------------------------------------------------------------------|-----------|
+| 1    | Platform build     | The Green environment is built with Kong Enterprise setup using the same version as the Blue environment |           |
+| 2    | Database setup     | A new Postgres DB is built and connected to the green environment                                     |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:45]:
```suggestion
| 2    | Database setup     | Build a new Postgres DB and connect it to the green environment                                     |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+| 3    | Regression Testing | A deployment is made into the new system and regression tested. The deployment is cleaned up when testing is completed |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:46]:
```suggestion
| 3    | Regression testing | Deploy your setup into the new system and regression test it. Clean up the deployment when testing is completed . |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+
+## Preperation
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:49]:
```suggestion
## Preparation
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+
+
+| Step | Name                          | Description                                                                                                 | Completed |
+|------|-------------------------------|-------------------------------------------------------------------------------------------------------------|-----------|
+| 1    | Change embargo                | A change embargo is placed on the environment preventing new deployments or configuration changes             |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:55]:
```suggestion
| 1    | Change embargo                | Place a change embargo on the old environment preventing new deployments or configuration changes             |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+| 2    | Blue environment backup       | A backup is taken on the blue environment's Cassandra database and placed into redundant storage              |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:56]:
```suggestion
| 2    | Blue environment backup       | Make a backup of the blue environment's Cassandra database and place it into redundant storage              |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+| 3    | Observability setup           | Dashboards are created to track the health of the new environment                                           |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:57]:
```suggestion
| 3    | Observability setup           | Create dashboards to track the health of the new environment                                           |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+| 4    | Go/No go checkpoint           | Go/No go decision point for the migration execution                                                         |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:58]:
```suggestion
| 4    | Go/No Go checkpoint           | Go/No Go decision point for the migration execution                                                         |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+
+
+## Execution
+
+The goal of this phase is to reproduce the Kong configuration in the blue environment and perform regression/smoke testing ahead of live traffic cut-over.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:64]:
```suggestion
The goal of this phase is to reproduce the {{site.base_gateway}} configuration in the blue environment and perform regression and smoke testing ahead of live traffic cut-over. 
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+
+## Execution
+
+The goal of this phase is to reproduce the Kong configuration in the blue environment and perform regression/smoke testing ahead of live traffic cut-over. 
+
+
+| Step | Name                  | Description                                                                                                           | Completed |
+|------|-----------------------|-----------------------------------------------------------------------------------------------------------------------|-----------|
+| 1    | Configuration dump    | - Execute the "deck dump" command against the blue environment, for each workspace to build a YAML representation of the Kong estate<br>- Execute "deck dump --rbac-resources-only" against the blue environment for each workspace to build a YAML representation of any RBAC resources created<br>- If using the Kong Developer Portal, run the "portal fetch" command to build a representation of the Dev Portal Assets for a workspace |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:69]:
```suggestion
| 1    | Configuration dump    | - Execute the `deck dump` command against the blue environment for each workspace to build a YAML representation of the Kong estate<br>- Execute `deck dump --rbac-resources-only` against the blue environment for each workspace to build a YAML representation of any RBAC resources created<br>- If using the Kong Dev Portal, run the `portal fetch` command via the Portal CLI to build a representation of the Dev Portal assets for a workspace |           |
```
"kong estate" is a term I've never heard before. Kong configuration? Environment? Entities maybe?

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+| 2    | Configuration sync    | - Change decK to work with the green environment, and execute "deck sync" to push the configuration to the green environment<br>- Change Portal CLI to work against the green environment, and execute "portal deploy" to push the Dev Portal configuration against the new environment |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:70]:
```suggestion
| 2    | Configuration sync    | - Change decK to work with the green environment, and execute `deck sync` to push the configuration to the green environment<br>- Change the Portal CLI to work against the green environment, and execute `portal deploy` to push the Dev Portal configuration against the new environment |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+| 3    | Regression Testing    | Execute regression tests against the green environment backed by PostGres                                               |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:71]:
```suggestion
| 3    | Regression testing    | Execute regression tests against the green environment backed by PostgreSQL                                             |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+| 4    | Go/No go checkpoint   | If all tests pass proceed with traffic cutover                                                                         |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:72]:
```suggestion
| 4    | Go/No Go checkpoint   | If all tests pass, proceed with traffic cutover                                                                         |           |
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+
+## Traffic cutover
+
+The purpose of this phase is to migrate traffic in a controlled manner, with a rapid fall back mechanism if a problem is detected. The wait times before increasing the canary split can be increased/decreased depending on risk appetite.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:77]:
```suggestion
The purpose of this phase is to migrate traffic in a controlled manner, with a rapid fallback mechanism if a problem is detected. You can increase or decrease the wait times before increasing the canary split depending on the risk tolerance.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+| 2    | Canary 5% of traffic to the new green environment      | Increase traffic split to 5% and monitor.                                                                                       |           |
+| 3    | Canary 10% of traffic to the new green environment     | Increase traffic split to 10% and monitor.                                                                                      |           |
+| 4    | Canary 50% of traffic to the new green environment     | Increase traffic split to 50% and monitor.                                                                                      |           |
+| 5    | Switch 100% of traffic to the new green environment    | Perform a full DNS switch over to point to the new green environment.                                                           |           |
+| 6    | Demise old platform                                    | After 24 hours when confidence is high, demise of the old platform.                                                              |           |
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:87]:
```suggestion
| 6    | Take down old platform                                    | After 24 hours when confidence is high, take down the old platform.                                                              |           |
```
Demise is a great word, but not one I would use in the docs 😅

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+
+
+
+## Limitations
+
+
+* If basic auth plugin is used, the passwords are hashed and encrypted in the database which a decK dump will not export and it is impossible to get the passwords out of the database in the cleartext. Migrating these credentials will have to be performed using admin api or Kong Manager.
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:95]:
```suggestion
* If using the Basic Authentication plugin, the passwords are hashed and encrypted in the database. A decK dump doesn't export these credentials, so it is impossible to get the passwords out of the database in the cleartext. You must migrate these credentials using the Admin Api or Kong Manager.
```

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+| 2    | Canary 5% of traffic to the new green environment      | Increase traffic split to 5% and monitor.                                                                                       |           |
+| 3    | Canary 10% of traffic to the new green environment     | Increase traffic split to 10% and monitor.                                                                                      |           |
+| 4    | Canary 50% of traffic to the new green environment     | Increase traffic split to 50% and monitor.                                                                                      |           |
+| 5    | Switch 100% of traffic to the new green environment    | Perform a full DNS switch over to point to the new green environment.                                                           |           |
+| 6    | Demise old platform                                    | After 24 hours when confidence is high, demise of the old platform.                                                              |           |
+
+
+
+
+## Limitations
```
Comment (lena-larionova) [inline app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:92]: Is limitations the right word? Can we call it "Next steps"?

**Review by lena-larionova — CHANGES_REQUESTED** [review]

File: app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md (hunk 8 above)
Comment (Guaris) [reply app/_src/gateway/production/running-kong/migrate-cassandra-to-postgres.md:36]: yes

**Review by lena-larionova — APPROVED** [review]
One minor edit, otherwise lgtm

**Comment by Guaris** [comment]
sent for review got a thumbs up. merging in. Ignoring vale for now.

**Pull request merged**